archunit/src/main/java/com/tngtech/archunit/library/freeze/TextFileBasedViolationStore.java

```diff
@@ -0,0 +1,235 @@
+            throw new StoreInitializationFailedException(String.format(
+                    "Creating new violation store is disabled (enable by configuration %s.%s=true)",
+                    ViolationStoreFactory.FREEZE_STORE_PROPERTY_NAME, ALLOW_STORE_CREATION_PROPERTY_NAME));
+        }
+        return rulesFile;
+    }
+
+    private void ensureExistence(File folder) {
+        checkState(folder.exists() && folder.isDirectory() || folder.mkdirs(), "Cannot create folder %s", folder.getAbsolutePath());
+    }
+
+    private void checkInitialization(boolean initializationSuccessful, String message, Object... args) {
+        if (!initializationSuccessful) {
+            throw new StoreInitializationFailedException(String.format(message, args));
+        }
+    }
+
+    @Override
+    public boolean contains(ArchRule rule) {
+        return storedRules.containsKey(rule.getDescription());
+    }
+
+    @Override
+    public void save(ArchRule rule, List<String> violations) {
+        log.debug("Storing evaluated rule '{}' with {} violations: {}", rule.getDescription(), violations.size(), violations);
+        if (!storeUpdateAllowed) {
+            throw new StoreUpdateFailedException(String.format(
+                    "Updating frozen violations is disabled (enable by configuration %s.%s=true)",
+                    ViolationStoreFactory.FREEZE_STORE_PROPERTY_NAME, ALLOW_STORE_UPDATE_PROPERTY_NAME));
+        }
+        String ruleFileName = ensureRuleFileName(rule);
+        write(violations, new File(storeFolder, ruleFileName));
+    }
+
+    private void write(List<String> violations, File ruleDetails) {
+        String updatedViolations = Joiner.on("\n").join(escape(violations));
+        try {
+            Files.write(ruleDetails.toPath(), updatedViolations.getBytes(UTF_8));
+        } catch (IOException e) {
+            throw new StoreUpdateFailedException(e);
+        }
+    }
+
+    private List<String> escape(List<String> violations) {
+        return replaceCharacter(violations, "\n", "\\\n");
+    }
+
+    private List<String> unescape(List<String> violations) {
+        return replaceCharacter(violations, "\\\n", "\n");
+    }
+
+    private List<String> replaceCharacter(List<String> violations, String characterToReplace, String replacement) {
+        return violations.stream().map(violation -> violation.replace(characterToReplace, replacement)).collect(toList());
+    }
+
+    private String ensureRuleFileName(ArchRule rule) {
+        String ruleDescription = rule.getDescription();
+
+        String ruleFileName;
+        if (storedRules.containsKey(ruleDescription)) {
+            ruleFileName = storedRules.getProperty(ruleDescription);
+            log.debug("Rule '{}' is already stored in file {}", ruleDescription, ruleFileName);
+        } else {
+            ruleFileName = newRuleFileName(ruleDescription);
+            log.debug("Assigning new ID {} to rule '{}'", ruleFileName, ruleDescription);
```

Comment:
In the new, more general version, we should probably rather log:
```suggestion
            log.debug("Assigning new file {} to rule '{}'", ruleFileName, ruleDescription);
```

Reply:
changed